src/server/pkg/hashtree/cache.go

```diff
@@ -4,35 +4,124 @@ import (
 	"fmt"
 	"io"
 	"reflect"
+	"sync"
 
-	lru "github.com/hashicorp/golang-lru"
+	"github.com/hashicorp/golang-lru/simplelru"
 	"github.com/pachyderm/pachyderm/src/server/pkg/localcache"
 	"github.com/sirupsen/logrus"
 )
 
+type nascentTree struct {
+	cond *sync.Cond
+	err  error
+}
+
 // Cache is an LRU cache for hashtrees.
 type Cache struct {
-	*lru.Cache
+	lruCache     simplelru.LRUCache
+	lock         sync.Mutex
+	syncEvict    *bool
+	nascentTrees map[interface{}]*nascentTree
+}
+
+func castValue(value interface{}) (HashTree, error) {
+	tree, ok := value.(HashTree)
+	if !ok {
+		return nil, fmt.Errorf("corrupted cache: expected HashTree, found: %v", reflect.TypeOf(value))
+	}
+	return tree, nil
+}
+
+func evict(value interface{}) {
+	if tree, err := castValue(value); err != nil {
+		logrus.Infof(err.Error())
+	} else {
+		if err := tree.Destroy(); err != nil {
+			logrus.Infof("failed to destroy hashtree: %v", err)
+		}
+	}
 }
 
 // NewCache creates a new cache.
 func NewCache(size int) (*Cache, error) {
-	c, err := lru.NewWithEvict(size, func(key interface{}, value interface{}) {
-		go func() {
-			tree, ok := value.(*dbHashTree)
-			if !ok {
-				logrus.Infof("non hashtree slice value of type: %v", reflect.TypeOf(value))
-				return
-			}
-			if err := tree.Destroy(); err != nil {
-				logrus.Infof("failed to destroy hashtree: %v", err)
-			}
-		}()
+	syncEvict := false
+	lruCache, err := simplelru.NewLRU(size, func(key interface{}, value interface{}) {
+		if syncEvict {
+			evict(value)
+		} else {
+			go evict(value)
+		}
 	})
 	if err != nil {
 		return nil, err
 	}
-	return &Cache{c}, nil
+	return &Cache{
+		lruCache:     lruCache,
+		syncEvict:    &syncEvict,
+		nascentTrees: make(map[interface{}]*nascentTree),
+	}, nil
+}
+
+// Close will synchronously evict all hashtrees from the cache, cleaning up any on-disk data.
+func (c *Cache) Close() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	*c.syncEvict = true
+	c.lruCache.Purge()
+}
+
+// GetOrAdd will atomically attempt to get the given HashTree from the cache,
+// and if it does not exist, it will generate the HashTree using the provided
+// function and store it in the cache before returning it. This is implemented
+// because the underlying LRU library does not provide a way to atomically
+// perform this operation without preconstructing the HashTree, which is a
+// relatively expensive operation.
+// If the generator panics, the cache will be left in an unreliable state.
+func (c *Cache) GetOrAdd(key interface{}, generator func() (HashTree, error)) (HashTree, error) {
```

Comment:
I really like using a callback to fill the cache, I think it might be a little cleaner to have one callback registered for the cache on construction, although it's not totally clear, because different callsites are passing different callbacks, so we'd need to make the single callback branch on whether the key is `"nil"` or not. Not sure if it's cleaner or not, and afaict there's only 2 callsites so it wouldn't really save any code. Up to you whether or not to implement this.